### PR TITLE
Pull Requests w/ Statuses

### PR DESCRIPTION
**Pull request by StanleyGoldman**
Initially we set out to implement both the Statuses API and the Checks API, but OAuth applications do not currently have access to the Checks API and there are no major projects making use of the Statuses API.

Since Statuses are simpler Check Runs, we can implement Statuses with the anticipation of adding Check Runs.

https://developer.github.com/v3/repos/statuses/

#### Known issues
- The icon will display the icon of the user that integrated the app, not the icon of the app itself
  - There is a bug in the GraphQL API where it is not possible to get the correct Application icon for Statuses. I thought I did everything for MEF to like me, but `VisualStudioBrowser` remains `null`.
  - We are waiting for API changes to make this data available
- [x] Trouble opening the Hyperlink from `PullRequestCheckRunView`
   https://github.com/github/VisualStudio/blob/2f3e739c126e52709a0649176d4a6591e193eb16/src/GitHub.VisualStudio/Views/GitHubPane/PullRequestCheckView.xaml.cs#L27-L34

#### New Metrics
- **NumberOfPRDetailsOpenInGitHub**: The number of times the "View on GitHub" link is clicked from a pull request detail view
- **NumberOfPRCheckStatusesOpenInGitHub**: The number of times a Status API "Details" link is clicked

#### TODO
- [x] Need designs for the Checks dropdown on the `PullRequestDetailsView` (@donokuda)

#### User Interfaces

image

image

### CODE REVIEW

**Comment by StanleyGoldman** [comment]
image

Nvm me..

**Comment by StanleyGoldman** [comment]
@meaghanlewis this is ready for you

**Comment by StanleyGoldman** [comment]
cc: @telliott27 I added two new metrics here

**Comment by meaghanlewis** [comment]
@StanleyGoldman this is looking and working great to me! ✨ 🎉 ⚡️ 

One thing I had a question about is the `avatar_url` (I believe that's what it's called). On dotcom, the icon displayed is the same as the service's logo:

github status

In Visual Studio, it's usually a person's avatar or another icon:
vs status

Was there a specific reason that the icons shown in VS are displayed this way? It might be better to mirror dotcom to make it more clear what service each status is coming from.

**Comment by StanleyGoldman** [comment]
@meaghanlewis this is a bug in the GraphQL API which is being addressed. I made mention of it in the body of this pull request.

The Icon being returned is the icon for the user that integrated the app, not the icon of the application itself.

You are right in that the contrast of what people are used to seeing vs what we are actually displaying, is confusing. Maybe we should remove the icons altogether until the GraphQL API can give us the correct icons.

**Comment by meaghanlewis** [comment]
Doh, sorry about that @StanleyGoldman! Thanks for opening up that issue to have further discussion about this.

File: src/GitHub.App/Models/PullRequestModel.cs [hunk 1]
```diff
@@ -125,6 +124,17 @@ public PullRequestStateEnum State
             }
         }
 
+        PullRequestChecksEnum checks;
```
Comment (StanleyGoldman) [inline src/GitHub.App/Models/PullRequestModel.cs]: asdf

File: src/GitHub.VisualStudio/Views/GitHubPane/PullRequestCheckView.xaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+             xmlns:views="clr-namespace:GitHub.VisualStudio.Views"
+             xmlns:ui="clr-namespace:GitHub.UI;assembly=GitHub.UI"
+             xmlns:vm="clr-namespace:GitHub.ViewModels.GitHubPane;assembly=GitHub.App"
+             xmlns:cache="clr-namespace:GitHub.UI.Helpers;assembly=GitHub.UI"
+             xmlns:sampleData="clr-namespace:GitHub.SampleData;assembly=GitHub.App"
+             xmlns:prop="clr-namespace:GitHub.VisualStudio.UI;assembly=GitHub.VisualStudio.UI"
+             mc:Ignorable="d" d:DesignWidth="600">
+
+    <d:DesignData.DataContext>
+        <vm:PullRequestCheckViewModel Title="continuous-integration/appveyor/branch"
```
Comment (grokys) [inline src/GitHub.VisualStudio/Views/GitHubPane/PullRequestCheckView.xaml]: Getting an error here:

```
The type "PullRequestCheckViewModel" does not include any accessible constructors.
```

Probably need a `PullRequestCheckViewModelDesigner` class.

File: src/GitHub.VisualStudio/Views/GitHubPane/PullRequestDetailView.xaml [hunk 3]
```diff
@@ -43,6 +43,8 @@
                 <ghfvs:SharedDictionaryManager Source="pack://application:,,,/GitHub.UI;component/Assets/Markdown.xaml" />
             </ResourceDictionary.MergedDictionaries>
 
+            <v:ViewLocator x:Key="viewLocator"/>
```
Comment (grokys) [inline src/GitHub.VisualStudio/Views/GitHubPane/PullRequestDetailView.xaml]: You shouldn't need a `ViewLocator` here. There's already one on the root `GitHubPaneView`.

File: src/GitHub.VisualStudio/Views/GitHubPane/PullRequestDetailView.xaml [hunk 4]
```diff
@@ -245,6 +247,25 @@
...
+                                      IsExpanded="True"
+                                      Margin="0 8 0 0"
+                                      ghfvs:ScrollingVerticalStackPanel.IsFixed="true">
+                    <ItemsControl ItemsSource="{Binding Checks}" Margin="0 4 12 4">
+                        <ItemsControl.ItemsPanel>
+                            <ItemsPanelTemplate>
+                                <StackPanel Grid.IsSharedSizeScope="True" />
+                            </ItemsPanelTemplate>
+                        </ItemsControl.ItemsPanel>
+                        <ItemsControl.ItemTemplate>
```
Comment (grokys) [inline src/GitHub.VisualStudio/Views/GitHubPane/PullRequestDetailView.xaml]: I don't think this `ItemTemplate` should be needed. The `ViewLocator` should automatically find the relevant view.

File: src/GitHub.InlineReviews/Services/PullRequestSessionService.cs [hunk 5]
```diff
@@ -353,6 +357,42 @@ public virtual async Task<PullRequestDetailModel> ReadPullRequestDetail(HostAddr
             return result;
         }
 
+        private async Task<LastCommitAdapter> GetPullRequestLastCommitAdapter(HostAddress address, string owner, string name, int number)
```
Comment (grokys) [inline src/GitHub.InlineReviews/Services/PullRequestSessionService.cs]: Nit: `private` not needed. Also private methods should go after `public`/`protected`.

File: src/GitHub.Exports/Models/StatusStateEnum.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿namespace GitHub.Models
+{
+    public enum StatusStateEnum
```
Comment (grokys) [inline src/GitHub.Exports/Models/StatusStateEnum.cs]: Again, the `Enum` suffix isn't needed.

File: src/GitHub.Exports.Reactive/ViewModels/GitHubPane/IPullRequestCheckViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        string Title { get; set; }
+        string Description { get; set; }
+        PullRequestCheckStatusEnum Status { get; set; }
+        Uri DetailsUrl { get; set; }
+        string AvatarUrl { get; set; }
+        BitmapImage Avatar { get; set; }
+        ReactiveCommand<object> OpenDetailsUrl { get; }
+    }
+
+    public enum PullRequestCheckStatusEnum
```
Comment (grokys) [inline src/GitHub.Exports.Reactive/ViewModels/GitHubPane/IPullRequestCheckViewModel.cs]: Again, the `Enum` suffix isn't necessary.

File: src/GitHub.App/ViewModels/GitHubPane/PullRequestCheckViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+        private void DoOpenDetailsUrl(object obj)
+        {
+            usageTracker.IncrementCounter(x => x.NumberOfPRCheckStatusesOpenInGitHub).Forget();
+        }
+
+        public string Title
+        {
+            get { return title; }
+            set { this.RaiseAndSetIfChanged(ref title, value); }
```
Comment (grokys) [inline src/GitHub.App/ViewModels/GitHubPane/PullRequestCheckViewModel.cs]: Can these properties change? If not then they shouldn't be settable and don't need to raise property changed events.

File: src/GitHub.App/ViewModels/GitHubPane/PullRequestListItemViewModel.cs [hunk 9]
```diff
@@ -31,6 +32,9 @@ public PullRequestListItemViewModel(PullRequestListItemModel model)
         /// <inheritdoc/>
         public IActorViewModel Author { get; protected set; }
 
+        /// <inheritdoc/>
```
Comment (grokys) [inline src/GitHub.App/ViewModels/GitHubPane/PullRequestListItemViewModel.cs]: Actually, I'm not sure why I made these properties `protected set` - I think that was a mistake... I think I initially wrote this as a base class for both issues and PRs and then decided not to do that.

File: src/GitHub.App/ViewModels/GitHubPane/PullRequestDetailViewModel.cs [hunk 10]
```diff
@@ -55,6 +55,8 @@ public sealed class PullRequestDetailViewModel : PanePageViewModelBase, IPullReq
         bool refreshOnActivate;
         Uri webUrl;
         IDisposable sessionSubscription;
+        IViewViewModelFactory viewViewModelFactory;
```
Comment (grokys) [inline src/GitHub.App/ViewModels/GitHubPane/PullRequestDetailViewModel.cs]: This should be `readonly` and placed at the top with the other services.

File: src/GitHub.App/Models/PullRequestModel.cs [hunk 11]
```diff
@@ -125,6 +124,17 @@ public PullRequestStateEnum State
...
+        public PullRequestChecksEnum Checks
```
[the remark this answers is not in the document]
Comment (StanleyGoldman) [reply src/GitHub.App/Models/PullRequestModel.cs]: I see...

File: src/GitHub.App/Services/PullRequestService.cs [hunk 12]
```diff
@@ -848,5 +897,15 @@ class ReviewAdapter
...
             public int Count => CommentCount + (!string.IsNullOrWhiteSpace(Body) ? 1 : 0);
         }
+
+        class LastCommitSummaryModel
+        {
+            public List<StatusSummaryModel> Statuses { get; set; }
+        }
+    }
+
+    public class StatusSummaryModel
```
[the remark this answers is not in the document]
Comment (StanleyGoldman) [reply src/GitHub.App/Services/PullRequestService.cs]: :+1:

File: src/GitHub.App/ViewModels/GitHubPane/PullRequestCheckViewModel.cs [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        private PullRequestCheckStatusEnum status;
+        private Uri detailsUrl;
+        private string avatarUrl;
+        private BitmapImage avatar;
+
+        public static IEnumerable<IPullRequestCheckViewModel> Build(IViewViewModelFactory viewViewModelFactory, PullRequestDetailModel pullRequest)
+        {
+            return pullRequest.Statuses?.Select(model =>
+            {
+                var statusStateEnum = model.State;
```
[the remark this answers is not in the document]
Comment (StanleyGoldman) [reply src/GitHub.App/ViewModels/GitHubPane/PullRequestCheckViewModel.cs]: Cool

File: src/GitHub.App/ViewModels/GitHubPane/PullRequestCheckViewModel.cs (hunk 8 above)
Comment (StanleyGoldman) [reply src/GitHub.App/ViewModels/GitHubPane/PullRequestCheckViewModel.cs]: Oh, but I see the point..

File: src/GitHub.App/Models/PullRequestModel.cs [hunk 14]
```diff
@@ -3,7 +3,6 @@
 using GitHub.Primitives;
 using GitHub.VisualStudio.Helpers;
 using System.Diagnostics;
-using System.Collections.Generic;
```
Comment (grokys) [inline src/GitHub.App/Models/PullRequestModel.cs]: Nit: Would be nice to completely remove the changes to these unchanged files to give a cleaner diff ;)

File: src/GitHub.App/SampleData/PullRequestCheckViewModelDesigner.cs [hunk 15]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Windows.Media.Imaging;
+using GitHub.ViewModels;
+using GitHub.ViewModels.GitHubPane;
+using ReactiveUI;
+
+namespace GitHub.App.SampleData
```
Comment (grokys) [inline src/GitHub.App/SampleData/PullRequestCheckViewModelDesigner.cs]: Namespace for this should simply be `GitHub.SampleData`.

File: src/GitHub.VisualStudio/Views/GitHubPane/PullRequestCheckView.xaml [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+             xmlns:ui="clr-namespace:GitHub.UI;assembly=GitHub.UI"
+             xmlns:vm="clr-namespace:GitHub.ViewModels.GitHubPane;assembly=GitHub.App"
+             xmlns:cache="clr-namespace:GitHub.UI.Helpers;assembly=GitHub.UI"
+             xmlns:sampleData="clr-namespace:GitHub.SampleData;assembly=GitHub.App"
+             xmlns:prop="clr-namespace:GitHub.VisualStudio.UI;assembly=GitHub.VisualStudio.UI"
+             xmlns:sampleData1="clr-namespace:GitHub.App.SampleData;assembly=GitHub.App"
+             mc:Ignorable="d" d:DesignWidth="600">
+
+    <d:DesignData.DataContext>
+        <sampleData1:PullRequestCheckViewModelDesigner />
```
Comment (grokys) [inline src/GitHub.VisualStudio/Views/GitHubPane/PullRequestCheckView.xaml]: If namespace for this class is changed as per previous comment you can use `ghfvs:` XML namespace.

File: src/GitHub.VisualStudio/Views/GitHubPane/PullRequestCheckView.xaml [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" 
+             xmlns:d="http://schemas.microsoft.com/expression/blend/2008" 
+             xmlns:local="clr-namespace:GitHub.VisualStudio.Views.GitHubPane"
+             xmlns:ghfvs="https://github.com/github/VisualStudio"
+             xmlns:c="clr-namespace:GitHub.VisualStudio.UI.Controls;assembly=GitHub.VisualStudio.UI"
+             xmlns:views="clr-namespace:GitHub.VisualStudio.Views"
+             xmlns:ui="clr-namespace:GitHub.UI;assembly=GitHub.UI"
+             xmlns:vm="clr-namespace:GitHub.ViewModels.GitHubPane;assembly=GitHub.App"
+             xmlns:cache="clr-namespace:GitHub.UI.Helpers;assembly=GitHub.UI"
+             xmlns:sampleData="clr-namespace:GitHub.SampleData;assembly=GitHub.App"
```
Comment (grokys) [inline src/GitHub.VisualStudio/Views/GitHubPane/PullRequestCheckView.xaml]: And this namespace isn't used.

File: src/GitHub.App/Services/ModelService.cs [hunk 18]
```diff
@@ -577,6 +577,7 @@ public PullRequestCacheItem(PullRequest pr)
             
             // Nullable for compatibility with old caches.
             public PullRequestStateEnum? State { get; set; }
+            public PullRequestChecksState? Checks { get; set; }
```
Comment (grokys) [inline src/GitHub.App/Services/ModelService.cs]: This isn't used any more.

File: src/GitHub.Exports/Models/StatusStateEnum.cs [hunk 19]
```diff
@@ -0,0 +1,11 @@
+﻿namespace GitHub.Models
+{
+    public enum StatusState
```
Comment (grokys) [inline src/GitHub.Exports/Models/StatusStateEnum.cs]: You renamed the enum but the filename is still `StatusStateEnum`.

File: src/GitHub.App/ViewModels/GitHubPane/PullRequestListItemViewModel.cs [hunk 20]
```diff
@@ -42,12 +46,12 @@ public bool IsCurrent
         }
 
         /// <inheritdoc/>
-        public int Number { get; protected set; }
+        public int Number { get; }
```
Comment (grokys) [inline src/GitHub.App/ViewModels/GitHubPane/PullRequestListItemViewModel.cs]: Thanks for fixing these :D

File: src/GitHub.VisualStudio/Views/GitHubPane/PullRequestCheckView.xaml [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+             xmlns:views="clr-namespace:GitHub.VisualStudio.Views"
+             xmlns:ui="clr-namespace:GitHub.UI;assembly=GitHub.UI"
+             xmlns:vm="clr-namespace:GitHub.ViewModels.GitHubPane;assembly=GitHub.App"
+             xmlns:cache="clr-namespace:GitHub.UI.Helpers;assembly=GitHub.UI"
+             xmlns:prop="clr-namespace:GitHub.VisualStudio.UI;assembly=GitHub.VisualStudio.UI"
+             xmlns:ghvs="clr-namespace:GitHub.SampleData;assembly=GitHub.App"
+             mc:Ignorable="d" d:DesignWidth="600">
+
+    <d:DesignData.DataContext>
+        <ghvs:PullRequestCheckViewModelDesigner />
```
Comment (grokys) [inline src/GitHub.VisualStudio/Views/GitHubPane/PullRequestCheckView.xaml]: I think you misunderstood me here: I meant you should be able to use the existing `ghfvs` namespace for this.

**Pull request merged**